patch-db/src/locker.rs

```diff
@@ -23,21 +24,109 @@ impl Locker {
             let mut locks_on_lease = vec![dummy_recv];
             let (_dummy_send, dummy_recv) = oneshot::channel();
             let mut cancellations = vec![dummy_recv];
+
+            let mut request_queue = VecDeque::<(Request, OrdSet<HandleId>)>::new();
             while let Some(action) =
                 get_action(&mut new_requests, &mut locks_on_lease, &mut cancellations).await
             {
+                // to prevent starvation we privilege the front of the queue and only allow requests that
+                // conflict with the request at the front to go through if they are requested by sessions that
+                // are *currently blocking* the front of the queue
+                fn process_new_req(
+                    hot_seat: Option<&(Request, OrdSet<HandleId>)>,
+                    req: Request,
+                    trie: &mut Trie,
+                    locks_on_lease: &mut Vec<oneshot::Receiver<LockInfo>>,
+                    request_queue: &mut VecDeque<(Request, OrdSet<HandleId>)>,
+                ) {
+                    match hot_seat {
+                        // hot seat conflicts and request session isn't in current blocking sessions
+                        // so we push it to the queue
+                        Some((hot_req, hot_blockers))
+                            if hot_req.lock_info.conflicts_with(&req.lock_info)
+                                && !hot_blockers.contains(&req.lock_info.handle_id) =>
+                        {
+                            request_queue.push_back((req, OrdSet::new()))
+                        }
+                        // otherwise we try and service it immediately, only pushing to the queue if it fails
+                        _ => match trie.try_lock(&req.lock_info) {
+                            Ok(()) => {
+                                let lease = req.complete();
+                                locks_on_lease.push(lease);
+                            }
+                            Err(blocking_sessions) => {
+                                request_queue.push_back((req, blocking_sessions))
+                            }
+                        },
+                    }
+                }
                 #[cfg(feature = "tracing")]
                 tracing::trace!("Locker Action: {:#?}", action);
                 match action {
                     Action::HandleRequest(mut req) => {
                         cancellations.extend(req.cancel.take());
-                        trie.handle_request(req, &mut locks_on_lease)
+                        let hot_seat = request_queue.pop_front();
+                        process_new_req(
+                            hot_seat.as_ref(),
+                            req,
+                            &mut trie,
+                            &mut locks_on_lease,
+                            &mut request_queue,
+                        );
+                        hot_seat.map_or((), |a| request_queue.push_front(a));
```

Comment:
```suggestion
                        if let Some(hot_seat) = hot_seat {
                            request_queue.push_front(hot_seat);
                        }
```
this is more idiomatic than performing a mutation in a map